### PR TITLE
Add support for extras in markers

### PR DESCRIPTION
**Pull request by BergLucas**
Resolves: [python-poetry#834](https://github.com/python-poetry/poetry/issues/834)

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

## Description

Hello dear maintainers, it's my first contribution to Poetry so I hope I have done the things right.

As you may have seen in the issue mentioned above, there's a feature that's been requested but hasn't yet been implemented. Since I also need this feature, I decided to try to implement it myself and contribute to the project.

In the issue, it is requested that Poetry implements a feature that allows differing sets of dependency extras in the `tool.poetry.extras` section. I've chosen to make a slightly different implementation that the ones that were proposed because they had some problems in my opinion but this implementation also fix their problems.

In the `pyproject.toml`, it would look like that :

```toml
[tool.poetry.dependencies]
dependency = { version = "...", optional = true }

[tool.poetry.extras]
feature-a = ["dependency[feature-a]"]
feature-b = ["dependency[feature-b]"]
```

The extras that are located in the `tool.poetry.extras` section are additional extras that are added to the ones in `tool.poetry.dependencies` section. So, if there were an extra named `feature-c` in `dependency`, installing `my_project[feature-a]` would install `dependency[feature-a,feature-c]`.

## Why this implementation ?

In the [2 implementations proposed in the issue](https://github.com/python-poetry/poetry/issues/834#issue-402992710), I thought that there was some problems.

In the one with aliases, it would required many changes to make it work and the dependencies would have been less clear since we would not be able to read the true package names directly.

In the one with the superset that can be filtered, I thought that it would be a good idea at first but on second thought, it was not a really good idea since it would create backward-incompatible changes. Indeed, if we have a `pyproject.toml` like the one below, how should it be interpreted ?

```toml
[tool.poetry.dependencies]
dependency = { version = "...", optional = true, extras = ["feature-a", "feature-b"] }

[tool.poetry.extras]
no-feature = ["dependency"]
```

In the current implementation, it would download both features but if we implemented the new feature like a superset that can be filtered, it would have been more logical that it download neither of the features.

That's the reason why I implemented the feature like additional extras that can be added in the `tool.poetry.extras` section. It is backward-compatible because the feature only applies when we add square brackets at the end of an extra and since it was not possible before, it does not break anything.

In addition, it involves only very minor modifications because we just need to add additional dependencies when we found an extra in the `tool.poetry.extras` section.

Finally, it also solves [another problem that someone mentioned in the issue](https://github.com/python-poetry/poetry/issues/834#issuecomment-1184186068). Indeed, with this implementation, it is possible to have optional extras by writing something like this :

```toml
[tool.poetry.dependencies]
dependency = { version = "..." }

[tool.poetry.extras]
feature = ["dependency[feature]"]
```

With this configuration, it would install `dependency[feature]` only when installing `my_project[feature]` and it would install `dependency` when installing `my_project`.

If you are wondering why I choose to add additional extras and not to override extras, it's only because it could create inconsistencies. For example, with the following configuration, installing `my_project[feature-c]` would not install `dependency[feature-a, feature-b]` and since it is in the required section, it is not normal :

```toml
[tool.poetry.dependencies]
dependency = { version = "...", extras=["feature-a", "feature-b"] }

[tool.poetry.extras]
feature-c = ["dependency[feature-c]"]
```

## Conclusion

Thanks for reading this far and sorry for writing so much. I really wanted to expose my ideas and it took more space than expected.

Please let me know if there are any changes that could be made, and thank you for creating this fantastic software.

### CODE REVIEW

**Review by radoering — CHANGES_REQUESTED** [review]
Thank you for the comprehensive issue description. Although I'm not completely sure about the design, you have highlighted why other approaches don't work so it might be the best solution for now. (Sooner or later we'll support PEP 621 anyway, which IMO will make it easier to specify extras.)

You've already considered a lot of things I might have forgotten, but there are still a few issues.

For example `poetry check` is still an issue because of [these lines](https://github.com/python-poetry/poetry-core/blob/f09d84a4c6ad6d1117998ff9d46949e1bd819f5b/src/poetry/core/factory.py#L415-L429). When running `poetry check` on your example pyproject.toml, you will get
```
Error: Cannot find dependency "transformers[torch,torch-vision]" for extra "torch" in main dependencies.
```

and so on. (See the inline comments for further issues.)

**Comment by dimbleby** [comment]
I'm not sure that new syntax is needed, I'd half expect something like this to work today:
```toml
foo = [
  { version = "^1.13.1" },
  { version = "^1.13.1", extras = ["their-extra", "their-other-extra"], markers = "extra == 'our-extra'" },
]
```
presumably it doesn't, but perhaps it wouldn't be so hard to make that work.

File: tests/test_factory.py [hunk 1]
```diff
@@ -494,3 +494,63 @@ def test_all_classifiers_unique_even_if_classifiers_is_duplicated() -> None:
...
+    assert [transformers_tensorflow] == extras[canonicalize_name("tf")]
+    assert transformers_tensorflow.pretty_constraint == "^4.30.2"
+    assert len(transformers_tensorflow.extras) == 1
+    assert len(transformers_tensorflow.in_extras) == 1
+
+    psycopg = find("psycopg", {"binary"})
+    assert all(psycopg not in extra for extra in extras.values())
+    assert psycopg.pretty_constraint == "^3.1.9"
+    assert len(psycopg.extras) == 1
+    assert len(psycopg.in_extras) == 0
```
[the remark this answers is not in the document]
Comment (radoering) [reply tests/test_factory.py]: That's python-poetry/poetry#2357

tldr: Optional dependencies that are not included in an extra make no sense and an error should be raised because that's not a valid configuration.

**Comment by radoering** [comment]
> I just think it's a bit confusing to specify an empty list in the extras section and then add the extras at the markers in the dependencies.

We could document that it's necessary or change it so that it's not necessary. No strong feelings so far. Either way, we should probably extend the [docs](https://python-poetry.org/docs/pyproject#extras) mentioning this alternative (more precise) method to define extras.

**Comment by dimbleby** [comment]
do `poetry install --extras tf` and `poetry install` both give the expected results in this example?  I guessed that something would go wrong with this, good news if not...

**Comment by BergLucas** [comment]
You were right @dimbleby, it doesn't install the extras when I run `poetry install --extras ...` even if it does create the right `METADATA` file using `pip wheel`.

I tested with `Poetry 1.6.1`.

**Comment by BergLucas** [comment]
I've tried to make the changes that you two suggested and it works for `poetry-core` but it does not when using `poetry`. I don't really know why but it is weird that Poetry does not use the data from the factory when installing the dependencies.

As for my changes, it does not work for `extra != '...'` at the moment and I'm not sure if I handle the `MultiMarker` and `MultiConstraint` classes correctly since it could potentially install dependencies only if multiples extras are enabled.

File: src/poetry/core/factory.py [hunk 2]
```diff
@@ -188,6 +229,18 @@ def configure_package(
                         dep.in_extras.append(extra_name)
                         package.extras[extra_name].append(dep)
 
+        for dep in package.requires:
+            extras = cls._get_extras_from_marker(dep.marker)
+            for extra in extras:
+                if extra not in dep.in_extras:
```
Comment (radoering) [inline src/poetry/core/factory.py:235]: How does this happen? Shouldn't `extra` be always be in `dep.in_extras`?

**Comment by radoering** [comment]
> I've tried to make the changes that you two suggested and it works for `poetry-core` but it does not when using `poetry`. I don't really know why but it is weird that Poetry does not use the data from the factory correctly when installing the dependencies.

Probably, one of these methods is not sufficient for this use case.

- https://github.com/python-poetry/poetry/blob/f5b022843660988daaf4153cf3058a33b389b4f6/src/poetry/installation/installer.py#L422
- https://github.com/python-poetry/poetry/blob/f5b022843660988daaf4153cf3058a33b389b4f6/src/poetry/installation/installer.py#L374
- https://github.com/python-poetry/poetry/blob/f5b022843660988daaf4153cf3058a33b389b4f6/src/poetry/installation/installer.py#L405

**Comment by Bibo-Joshi** [comment]
Hi. Just wanted to add two more observations:

1. pip seems to be happy enough with the produced metadata. `pip install .` and `pip install .[extra]` work as expected. only tested with simple `extra = '…'` markers and tbh I don't quite get yet if the "extra != , and, or, in" cases are poetry specific or not :D.
2. With poetry 1.7.1, I observe the contrary of 
  > But when I run poetry install -E extra-binary, it tries to install everything:

  i.e. poetry install none of the extra requirements

**Comment by GOGKI** [comment]
Hi, will this PR be merged and closed? Is there an ETA?

**Comment by just-maiyak** [comment]
Hello ! Will this be looked into ?

**Comment by radoering** [comment]
I am closing this PR because it has obviously stalled and circumstances have changed in the meantime: `tool.poetry.extras` will be deprecated in favor of `project.optional-dependencies` in the next version so that it does not make sense to extend it.

In case you are looking for a solution for python-poetry/poetry#834, depending on what exactly you want to do this might already work in Poetry's main branch by specifying extras in `project.optional-dependencies` (if there are no conflicts between extras) or your best bet is python-poetry/poetry#9553 (if there are conflicts).

**Comment by BergLucas** [comment]
It works well in Poetry v2, thanks for your work!
